detection to gather all azure app ids, and process via loop for sharepoint and exo investigation

crash0ver1d3 opened the pull request:
used get-credential to store creds to variable to be passed to the -credential parameter of connection commandlets
variable gets set to $null at the end of the script.

added detection to gather all azure app ids, and feed to loop for investigation of all appIds
-sub feature of creating a folder for app investigations, and then creating sub-folders named with the app displayname.
mailitems and fileitems export is dumped into directory with corresponding displayname

# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->
The goal is to reduce the manual interaction of this script overall. This was accomplished by reducing the number of times 
credentials would have to be entered by utilizing get-credential commandlet, and storing creds to variable.
Also, the option to investigate all Azure App IDs was added. This gathers all of the appids that are available
and performs the UAL query respectively.

## 💭 Motivation and Context ##

<!-- Why is this change required? -->
This change is required, because the current state of the code only allows for single appID investigation
By adding the option to investigate all AppIds, this gives more flexibility and reduces workload to correlate actions to investigate.
This will greatly increase accuracy, and give DFIR an option to look for all appid interaction, and not just a single appid at every script execution.
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
I used an AzureCloud and 0365Default environment to validate the proper function of the script, and correlation of data, after making my changes. Anyone with access to an Azure and 0365 tenant can test these changes.
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
Before modifying the code, I make an archive of the results that the script generates, prior to edits.
After passing my unit tests for my code, I verified that the content matches the offline archive, plus the addition of my data.
I do have an E5 license, so I was able to fully text the MailBoxItemsAccessed.
## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [ ] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
* [ ] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [ ] All new and existing tests pass.

crash0ver1d3 committed:
added detection to gather all azure app ids, and feed to loop for investigation of all appIds

used get-credential to store creds to variable to be passed to the -credential parameter of connection commandlets
variable gets set to $null at the end of the script.

added detection to gather all azure app ids, and feed to loop for investigation of all appIds
-sub feature of creating a folder for app investigations, and then creating sub-folders named with the app displayname.
mailitmes export is dumped into directory with corresponding displayname

crash0ver1d3 committed:
Update Sparrow.ps1

crash0ver1d3 committed:
corrected varible name

$InvestigationExportDir was corrected to $InvestigationMailExportDir to resolve issue where sub-diretories by App DisplayName were not getting created.

crash0ver1d3 committed:
corrected parameter binding value

In the Get-AzureSPAppRoles function, defined $Credential parameter as $false, to match other functions. This allows someone to not use the $creds=get-credential, and prompt for authentication at each connection instances.

DeemOnSecurity commented:
Overall like the implementation. I apologize for the delay! Please address the review comments and we will merge, if not I will try to implement some time next week.

crash0ver1d3 commented:
Thanks Will. I will review the comments and update. I am doing a
training class this week, so it will most likely be this weekend.

Glad you like the update. Hopefully this will help other people, which is
my sole motivation.

The script overall is very nicely done. I like using the hashtable to
create an array of desired values. Very nice.

On Tue, Jan 19, 2021 at 10:33 AM Will Deem <notifications@github.com> wrote:

> Overall like the implementation. I apologize for the delay! Please address
> the review comments and we will merge, if not I will try to implement some
> time next week.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cisagov/Sparrow/pull/45#issuecomment-762961774>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJQ5GZRGNC3HXOZNMI7F3STS2WX5FANCNFSM4V42KISQ>
> .
>


-- 
Regards
Blake Regan

genericdevname commented:
Would you mind editing the loop so it looks at service principals that are identified as applications? Should be a simple fix--something like `Get-AzureADServicePrincipal -All $true | Where-Object {$_.ServicePrincipalType -eq "Application"}`

genericdevname commented on Sparrow.ps1:
```diff
@@ -143,12 +151,29 @@ Function Get-UALData {
+    }
+    If ($AppIdQuestion_GetAll -eq "Yes"){
+        write-Host "Gathering Azure Application IDs..."
+        $AzureAppIds=Get-AzureADApplication -All:$true
```
Recommend using `Get-AzureADServicePrincipal -All $true | Where-Object {$_.ServicePrincipalType -eq "Application"}` instead

crash0ver1d3 replied:
The requested change has been added. Also added comments to the file creation logic.

crash0ver1d3 committed:
added  CISA review change request, and comments

Tested the update for accuracy. Confirmed succcessful execution and results generated.

crash0ver1d3 reviewed (commented):
Is there anything else that you need from me before merging this pull request?

genericdevname commented:
We made a push for the SAML logic so we're currently working on updating your pull request with the newest logic along with some minor changes.

crash0ver1d3 commented:
Sounds good. If you need me to test my PR, with the updated logic, let me
know.

On Mon, Feb 8, 2021 at 8:44 AM genericdevname <notifications@github.com>
wrote:

> We made a push for the SAML logic so we're currently working on updating
> your pull request with the newest logic along with some minor changes.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cisagov/Sparrow/pull/45#issuecomment-775198488>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJQ5GZQ6GKTGVZUVUOIMEHDS572E3ANCNFSM4V42KISQ>
> .
>


-- 
Regards
Blake Regan

genericdevname committed:
Included new changes

genericdevname commented:
@crash0ver1d3 Can you do a quick test? I have tested the new logic, but wanted to make sure your changes were accurately captured. Thanks!

Had to remove the credentials parameter as it did not work with modern authentication and MFA.

crash0ver1d3 commented:
I have done a sanity check, and the script does generate results as expected. You saved a PR that I was going to submit, about not creating folders and csv for appId with no results. Nicely done.
Noting, I did have to enter my credentials 5 different times. I understand the need to use ADAL and MFA, but this makes this script not able to be run in an automated fashion in this state.

genericdevname committed:
Merge branch 'develop' into Sparrow_Investigate_All_AppIds
